velox/type/Filter.cpp

```diff
@@ -458,6 +472,59 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      bool bothNanAllowed = nanAllowed_ && multiRangeOther->nanAllowed_;
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+            case FilterKind::kIsNull:
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        return merged[0]->clone(bothNullAllowed);
+      } else {
+        return std::unique_ptr<Filter>(
+            new MultiRange(std::move(merged), bothNullAllowed, bothNanAllowed));
+      }
+    }
+      /*
+       * Rules of MultiRange with IsNull/IsNotNull
+       * MultiRange(nullAllowed=true) AND IS NULL => IS NULL
+       * MultiRange(nullAllowed=true) AND IS NOT NULL =>
+       * MultiRange(nullAllowed=false) MultiRange(nullAllowed=false) AND IS NULL
+       * => ALWAYS FALSE MultiRange(nullAllowed=false) AND IS NOT NULL =>
+       * MultiRange(nullAllowed=false)
+       */
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull: {
+      if (this->nullAllowed_) {
+        return this->clone(/*nullAllowed=*/false);
```

Comment:
@mbasmanova I dont understand why
MultiRange(nullAllowed=true) AND IS NOT NULL => MultiRange(nullAllowed=false)

Reply:
A value passes F1 AND F2 only if it passes both F1 and F2. When F2 = IS NOT NULL, any value other than null passes, e.g. null does not pass. However, null passes MultiRange(nullAllowed=true). To exclude null we simply change to MultiRange(nullAllowed=false)